docs/usage/configuration/config_documentation.md

```diff
@@ -462,6 +462,21 @@ See the docs [request log format](../administration/request_log.md).
 * `additional_resources`: Only valid for an 'http' listener. A map of
    additional endpoints which should be loaded via dynamic modules.
 
+  _Added in Synapse 1.86.0_
+
+* UNIX socket support:
+  * `path`: A path and filename for a UNIX socket. Ensure it is in a readable/writeable
+     directory and is pre-existing(the directory will not be created). No default
+    * **Note**: Declaring a `path` is not compatible while also having a `port`
+       configuration for the same `listener`
+    * **Note**: `metrics` and `manhole` are not supported as `types` on UNIX socket `listeners`
+    * The `x_forwarded` option will default to true and may be left out.
+    * Other options that would not make sense to use with a UNIX socket, such as 
+    `bind_address[es]` and `tls` will be ignored and can be removed.
+  * `mode`: The file permissions to set on the UNIX socket. Defaults to `666`
```

Comment:
Why does the file `mode` matter to customize here? Seems like we should just use a sane default.

What's the least restrictive permissions we can use here?

Reply:
It felt to me that some who value the most locked down version of everything would appreciate being able to customize this. `666` is probably be the best for a default. However,(and I use Complement as my context) I believe anything more restrictive prohibits usage by nginx for the reverse proxy as it is a different 'user'(and would therefore classify as 'Everyone/No-one' which is what the last digit controls). As a visual test of what happens if I change the default to a more restrictive `660`:
```log
nginx | 2023/06/15 20:37:50 [crit] 74#74: *171 connect() to unix:/run/main_public.sock failed (13: Permission denied) while connecting to upstream, client: 172.17.5.1, server: localhost, request: "GET /_matrix/client/versions HTTP/1.1", upstream: "http://unix:/run/main_public.sock:/_matrix/client/versions", host: "127.0.0.1:32788"
```
floods the logs

In theory, you could set up a special user for both your reverse proxy and Synapse so they would share permissions. But that is an exercise for a security professional.

TLDR: Like a TCP port, you want anything and everything to be able to read/write it, authentication should be handled by the individual process on either side.

Reply:
It seems like `666` is the sane default :+1: 

And the only other case really is `660` if they share the same user.

Mostly asking this question to avoid extra unnecessary config. I feel like we could get away without the config and just use `666`. People could always pre-create the socket file with their custom permissions.

Reply:
Unfortunately, pre-creating a socket file doesn't work(with or without custom permissions).
<details><summary>The logs when that happens</summary>

```log
synapse_main | 2023-06-16 21:44:48,715 - synapse.app._base - 269 - CRITICAL - sentinel - Error during startup
synapse_main | Traceback (most recent call last):
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/unix.py", line 411, in startListening
synapse_main |     skt.bind(self.port)
synapse_main | OSError: [Errno 98] Address already in use
synapse_main |
synapse_main | During handling of the above exception, another exception occurred:
synapse_main |
synapse_main | Traceback (most recent call last):
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 254, in wrapper
synapse_main |     await cb(*args, **kwargs)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 363, in start
synapse_main |     await _base.start(hs)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 588, in start
synapse_main |     hs.start_listening()
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 251, in start_listening
synapse_main |     self._listener_http(self.config, listener)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 141, in _listener_http
synapse_main |     ports = listen_http(
synapse_main |             ^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 435, in listen_http
synapse_main |     ports = listen_unix(
synapse_main |             ^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 384, in listen_unix
synapse_main |     cast(Port, reactor.listenUNIX(path, factory, backlog, mode, wantPID))
synapse_main |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/posixbase.py", line 270, in listenUNIX
synapse_main |     p.startListening()
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/unix.py", line 413, in startListening
synapse_main |     raise error.CannotListenError(None, self.port, le)
synapse_main | twisted.internet.error.CannotListenError: Couldn't listen on any:b'/run/main_public.sock': [Errno 98] Address already in use.
synapse_main | Error during startup:
synapse_main | Traceback (most recent call last):
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/unix.py", line 411, in startListening
synapse_main |     skt.bind(self.port)
synapse_main | OSError: [Errno 98] Address already in use
synapse_main |
synapse_main | During handling of the above exception, another exception occurred:
synapse_main |
synapse_main | Traceback (most recent call last):
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 254, in wrapper
synapse_main |     await cb(*args, **kwargs)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 363, in start
synapse_main |     await _base.start(hs)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 588, in start
synapse_main |     hs.start_listening()
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 251, in start_listening
synapse_main |     self._listener_http(self.config, listener)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 141, in _listener_http
synapse_main |     ports = listen_http(
synapse_main |             ^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 435, in listen_http
synapse_main |     ports = listen_unix(
synapse_main |             ^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 384, in listen_unix
synapse_main |     cast(Port, reactor.listenUNIX(path, factory, backlog, mode, wantPID))
synapse_main |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/posixbase.py", line 270, in listenUNIX
synapse_main |     p.startListening()
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/unix.py", line 413, in startListening
synapse_main |     raise error.CannotListenError(None, self.port, le)
synapse_main | twisted.internet.error.CannotListenError: Couldn't listen on any:b'/run/main_public.sock': [Errno 98] Address already in use.
```

</details>

Reply:
Sounds like there is some extra locking or detail of the socket file content itself. Perhaps some command like this is necessary to create one: https://serverfault.com/a/914572/182342

Feels like it's possible to pre-create a socket file since the socket file stays behind after the app quits and starts again where it does work.

Reply:
I run into the same error when using that command mentioned:

```
$ python -c "import socket as s; sock = s.socket(s.AF_UNIX); sock.bind('/tmp/test.sock')"
$ python -c "import socket as s; sock = s.socket(s.AF_UNIX); sock.bind('/tmp/test.sock')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
OSError: [Errno 98] Address already in use
```

Not sure how the socket file normally gets cleaned up, etc so all of this works. But I guess `mode` is useful :shrug:

Reply:
The lock file contains the PID of the process that started it...it shouldn't stick around after the process exits, just like the socket file. I believe Twisted cleans it up automatically. It feels like we're branching a little into the theoretical with this...but yes it's possible to create the socket file externally, however that isn't behavior we should support. We want the server to create and clean up it's own socket like it does now. Creating an socket externally would require adding extra binding and other such stuff which complicates the matter a great deal.

That page you linked, the top answer defines the behavior we want(I think).

Reply:
> It feels like we're branching a little into the theoretical with this...

I think it's mainly because I'm not familiar and just working on recall with the few times I've used them in other systems.

Using standard patterns makes sense and I think I have a better understanding now :+1:

Reply:
And TLDR is `mode` is fine :fast_forward: